File: Intersection Point in Y Shaped Linked Lists.cpp
class Solution {
public:
    int intersectPoint(Node* head1, Node* head2) {
        if(!head1 or !head2) return -1;
        
        Node* curr = head1;
        while (curr->next) curr = curr->next;
        curr->next = head2;
        
        Node *slow = head1, *fast = head1;
        while (fast and fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            
            if (slow == fast) {
                fast = head1;
                while (slow != fast) {
                    slow = slow->next;
                    fast = fast->next;
                }
                return slow->data;
            }
        }
        return -1;
    }
};

File: Intersection Point in Y Shaped Linked Lists.java
class Intersect

{

  int getLength(Node head)

    {

        int l=0;

        while(head!=null)

        {

            head=head.next;

            l++;

        }

        return l;

    }

    

 int intersectPoint(Node head1, Node head2)

 {   

     

     int l1=getLength(head1);

     int l2=getLength(head2);

 

 

        while(l1>l2)

        {

            head1=head1.next;

            l1--;

        }

            

     while(l2>l1)

     {

         head2=head2.next;

         l2--;

     }

     

     

     while(head1!=null && head2!=null)

     {

         if(head1.equals(head2))

             return head1.data;

         head1=head1.next;

         head2=head2.next;

     }

     

  

     

     return -1;

 }

}
